### PR TITLE
feat(ee): RFC 8693 token exchange and per-organization API clients

### PR DESCRIPTION
**Pull request by harry-rhesis**
## Purpose

Adds an RFC 8693 OAuth 2.0 Token Exchange endpoint and the supporting per-organization OAuth client primitive (the EE `API_CLIENTS` feature). External integrations trade a Keycloak access token for a Rhesis JWT minted on behalf of the original user, without ever holding a long-lived Rhesis session.

The endpoint is the missing piece that turns Rhesis SSO from "humans log in through a browser" into "machines log in too, end-to-end attested by the org's own IdP".

### New endpoint and feature surface

- **`POST /auth/token-exchange`** — RFC 8693 token exchange. Trades an OIDC subject access token (validated against the calling org's `SSOConfig`) for a Rhesis JWT carrying `azp` / `aud` / `scope` / `jti` / `epoch` claims. The orchestrator (`ee/.../sso/token_exchange/exchange.py`) runs ten ordered security checks; reordering changes either the timing-oracle surface or skips a check entirely.
- **`POST /auth/refresh` (extended)** — refresh tokens minted by token exchange carry a `client_id` binding that the refresh request must re-prove via HTTP Basic. Core delegates to an EE-registered hook so the unbound-token path keeps Community-only behaviour.
- **EE feature `API_CLIENTS`** — gated on `FeatureRegistry`, with a runtime check that composes onto SSO (token exchange needs `SSOConfig.issuer_url`).
- **Org-scoped CRUD** — `/organizations/{id}/auth-clients` (create / list / get / rotate / disable / enable / delete). Plaintext secret is shown exactly once at create / rotate; the at-rest column is Fernet-encrypted on top of a versioned `sha256:` hash.

### New tables and columns

- `auth_client` (new, per-org) — secret hash, expected `azp`, expected audience, allowed scopes, default scope, `token_epoch` for coarse revocation, `disabled` soft-disable. Partial-unique index on `(organization_id, client_id) WHERE deleted_at IS NULL` so soft-deleted rows can be re-created without losing audit history.
- `refresh_token` — added nullable `client_id` and `scope` columns. Existing UI/SSO refresh tokens leave them NULL and behave unchanged.

### Frontend

- New EE module `ee/frontend/src/api-clients/` — registers an "API Clients" section in org settings (gated by `<FeatureGate feature={FeatureName.API_CLIENTS}>`), with create / list / rotate / delete dialogs and a one-shot secret-display dialog.
- Org settings page restructured to render registered EE sections inline.

### Security model (full breakdown in `docs/content/contribute/backend/api-clients.mdx`)

The orchestrator is hardened against a documented threat model: timing oracles, `alg=none` confusion, JWKS exhaustion, cross-org mint, sibling-integration token replay (A3 in the docs), refresh-token theft, and audit-side leakage of raw email / tokens.

Highlights:
- **Org-scoped client authentication** — `authenticate_client(db, org_id, client_id, secret)`. Two tenants may legitimately share the same `client_id`; a lookup keyed on `client_id` alone would lock one of them out.
- **Constant-time auth** — dummy-hash on missing client, `hmac.compare_digest` on every comparison; a CI test (`tests/backend/test_secret_equality.py`) bans `==` on hash-shaped strings repo-wide.
- **Subject-token replay protection** — Redis `SET NX EX` keyed by `(issuer_hash, jti)` so two IdPs that legitimately mint the same `jti` can't DoS each other.
- **`azp` + `aud` binding** — both are required at AuthClient create time; without `aud`, IdPs that share `azp` across siblings (Keycloak service-account flows) would let A3 succeed.
- **Coarse revocation via `token_epoch`** — bumped to `now() + 1s` on rotate to close the integer-second `iat >= epoch` race window. `verify_jwt_token` enforces `iat >= epoch` at every request.
- **Refresh hardening** — `/auth/refresh` checks `user.is_active`; the EE minter re-asserts `auth_client.organization_id == user.organization_id`; every 401 returns a uniform detail string so it cannot serve as an oracle.
- **Per-client + per-org rate limits** — wired post-parse via a new `hit_post_parse_limit` helper (the slowapi `@limit` decorator runs before the body is read, so it can only key on IP).
- **Stream-based body cap** — `/auth/token-exchange` reads the body via a counted streamer that holds even when `Content-Length` is missing (chunked transfer / HTTP/2).
- **Cache hardening** — `TokenEndpointCacheHeadersMiddleware` stamps `Cache-Control: no-store`, `Pragma: no-cache`, `X-Content-Type-Options: nosniff` on every response from `/auth/token-exchange` and `/auth/refresh` regardless of status.
- **Structured audit** — `AUTH_CLIENT_*` lifecycle events and `TOKEN_EXCHANGE_*` outcome events. Email is HMAC-SHA256 hashed (`AUDIT_HASH_KEY`); raw tokens / secrets / email never appear. EE bootstrap refuses to start in non-dev without `AUDIT_HASH_KEY` set.
- **Subject-token validation** — algorithm allowlist (`RS256`, `ES256`, `EdDSA`), header preflight (`alg` + `kid` required), strict issuer match (no `startswith`), JWKS rotation with per-issuer cooldown.
- **Async deadlock fix** — `_get_jwks` resolves OIDC metadata _outside_ the JWKS cache lock; the previous nested-acquire would deadlock on a cold cache.

### Documentation

- `docs/content/contribute/backend/api-clients.mdx` — full feature reference: data flow, security checks in order, refresh flow fan-out, coarse revocation, audit log fields, and how to add a new EE feature on the same primitives.

## Additional context

- The work is split across **12 commits** on this branch in dependency order: prerequisite helpers first (rate-limit, replay namespace), then the bulk feature commit, then the security-review hardening, then dead-code cleanup, then the partial-unique migration, then tests. Each commit message explains the security finding it addresses.
- A staff-engineer security review was performed mid-stream; 12 of 13 findings are addressed in this PR. The remaining one (P0 — replace the placeholder `is_superuser`/owner gate with a real RBAC primitive on `/organizations/{id}/auth-clients/*` and `/organizations/{id}/sso/*`) is being handled in a follow-up PR.
- The `_require_org_admin*` placeholders in the EE routers behave today as "any authenticated org member can administer" — that's the gap the follow-up RBAC PR closes. The data-plane (`/auth/token-exchange`, `/auth/refresh`) is fully gated by client credentials and is not affected.

## Migrations

Three new alembic revisions, applied in order:

1. `c7f4d9b2e1a3` — create `auth_client` table.
2. `d8a5e0f3c4b2` — add nullable `client_id` and `scope` to `refresh_token`.
3. `d8b3a1f2c4e5` — convert `(organization_id, client_id)` UniqueConstraint to a partial unique index `WHERE deleted_at IS NULL`; same predicate added to the `(organization_id, name)` index.

All three are idempotent.

## Configuration

New env vars:

| Variable | When required | Purpose |
|---|---|---|
| `AUDIT_HASH_KEY` | Required in production | HMAC key for `hashed_email` in the audit stream. EE bootstrap refuses to start without it in non-dev environments. |
| `DB_ENCRYPTION_KEY` | Already required for SSO | Reused for `auth_client.client_secret_hash` Fernet encryption. |

## Testing

Backend test suite covering this PR (run from `apps/backend/`):

```bash
RHESIS_SKIP_MIGRATIONS=1 uv run pytest \
  ../../tests/backend/ee/ \
  ../../tests/backend/auth/test_used_token_store.py \
  ../../tests/backend/auth/test_session_utils.py \
  ../../tests/backend/auth/test_slug_validation.py \
  ../../tests/backend/utils/test_rate_limit.py \
  ../../tests/backend/test_secret_equality.py \
  ../../tests/backend/test_ee_boundary.py
```

Result on this branch: **218 passed**.

Coverage areas:
- `tests/backend/ee/sso/token_exchange/test_exchange_helpers.py` — orchestrator helper edge cases.
- `tests/backend/ee/sso/token_exchange/test_router_status_mapping.py` — RFC 6749 error code -> HTTP status mapping.
- `tests/backend/ee/api_clients/test_clients.py` — `authenticate_client` org-scoping (regression for the multi-tenant collision bug); secret-hash determinism; dummy-hash on miss.
- `tests/backend/ee/api_clients/test_schemas.py` — `expected_subject_audience` is required at the Pydantic layer.
- `tests/backend/ee/api_clients/test_refresh_minter.py` — `_strip_offline_access` semantics (substring guard).
- `tests/backend/ee/api_clients/test_cache_headers.py` — `Cache-Control: no-store` on every status code.
- `tests/backend/ee/api_clients/test_refresh_client_hook.py` — single-slot semantics for the EE refresh-minter hook.
- `tests/backend/auth/test_used_token_store.py` — namespace argument changes the Redis key shape (regression for cross-IdP collision).
- `tests/backend/utils/test_rate_limit.py` — `hit_post_parse_limit` budget enforcement, per-(namespace, key) isolation.
- `tests/backend/test_secret_equality.py` — repo-wide CI ban on `==` for hash-shaped strings.
- `tests/backend/test_ee_boundary.py` — extended to check core/EE import direction is not violated by the new files.

End-to-end smoke test was performed locally against a small FastAPI demo app that:
1. Logs in to Keycloak as a user.
2. Calls `POST /auth/token-exchange` with the Keycloak access token + AuthClient credentials.
3. Calls `GET /test_sets/` with the resulting Rhesis JWT.
4. Calls `POST /auth/refresh` with HTTP Basic to rotate the refresh token.

The demo app itself is not part of this PR.

### CODE REVIEW

File: ee/backend/src/rhesis/backend/ee/api_clients/router.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+    if not user:
+        raise HTTPException(
+            status_code=status.HTTP_401_UNAUTHORIZED,
+            detail="Authentication required",
+        )
+
+    if getattr(user, "is_superuser", False):
+        return user
+
+    if str(user.organization_id) != org_id:
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply ee/backend/src/rhesis/backend/ee/api_clients/router.py:114]: Acknowledged, and thanks for the catch.

`_require_org_admin_for` is a placeholder that mirrors the SSO router's existing `_require_org_admin` -- both currently behave as "any authenticated org member can administer". Replacing this with a real RBAC primitive (admin / owner privilege check) is being handled in a **separate follow-up PR** that covers both `/organizations/{id}/auth-clients/*` and `/organizations/{id}/sso/*` together, so the permission story for SSO and API Clients lands in one consistent change rather than two staggered ones.

Anything in this PR regarding authorization / admin access on the management routes is deferred to that follow-up by design.

The data plane on this PR (`/auth/token-exchange`, `/auth/refresh`) is unaffected: it is gated by AuthClient credentials, the org-scoped feature gate, and (for refresh) HTTP Basic + the `client_id` binding on the refresh row. No part of the data plane goes through `_require_org_admin_for`.

File: ee/backend/src/rhesis/backend/ee/sso/token_exchange/router.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+    summary="RFC 8693 token exchange",
+)
+@limiter.limit(TOKEN_EXCHANGE_PER_IP_RATE_LIMIT)
+async def token_exchange(
+    request: Request,
+    db: Session = Depends(get_db_session),
+):
+    """Exchange a Keycloak (or other OIDC) access token for a Rhesis JWT.
+
+    The endpoint is feature-gated via ``FeatureName.API_CLIENTS``;
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply ee/backend/src/rhesis/backend/ee/sso/token_exchange/router.py:318]: Fixed in 1a54626 -- the gate is enforced inside `run_token_exchange` (after org resolution, before client auth) so the router-level docstring now matches runtime behavior. Returns RFC 8693 `invalid_target` / `feature_unavailable` and the same code path is covered by `tests/backend/ee/sso/token_exchange/test_feature_gating.py`.

File: ee/backend/src/rhesis/backend/ee/sso/token_exchange/exchange.py [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+            subject_token_jti=kwargs.get("jti"),
+            email=kwargs.get("email"),
+            scope=kwargs.get("scope"),
+            reason_code=reason_code,
+            ip=audit_ip,
+            user_agent=audit_ua,
+        )
+        raise TokenExchangeError(error, reason_code, http_status=http)
+
+    # ---- Step 2: org resolution ------------------------------------------
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply ee/backend/src/rhesis/backend/ee/sso/token_exchange/exchange.py:245]: Fixed in 1a54626 -- `run_token_exchange` now calls `FeatureRegistry.is_available(FeatureName.API_CLIENTS, org)` immediately after org resolution and bails out with RFC 8693 `invalid_target` / `feature_unavailable` (no enumeration leak) before any client-secret comparison. Regression coverage in `tests/backend/ee/sso/token_exchange/test_feature_gating.py` (`test_denies_when_feature_unavailable` + `test_proceeds_past_feature_check_when_available`).

File: apps/backend/src/rhesis/backend/app/auth/token_utils.py [hunk 4]
```diff
@@ -118,6 +223,28 @@ def verify_jwt_token(token: str, secret_key: str, algorithm: str = ALGORITHM) ->
             logger.warning(f"Invalid token type: {payload.get('type')}")
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/app/auth/token_utils.py:236]: Fixed in 1a54626 -- enforced on both sides:

- `verify_jwt_token` rejects any JWT that carries `azp` without `epoch` (raises `JWTError("Token missing required epoch claim")`), so a buggy mint path can no longer produce silently non-revocable client-bound tokens.
- `create_session_token` has a matching guard that raises `ValueError` if a caller passes `azp` without `epoch`, catching the issue at source rather than at verification.

Tests: `test_verify_jwt_token_rejects_azp_without_epoch` and `test_verify_jwt_token_accepts_azp_with_epoch` in `tests/backend/utils/test_auth_utils.py`.
